lib/vdsm/virt/thinp.py

```diff
@@ -211,99 +223,142 @@ def on_block_threshold(self, target, path, threshold, excess):
+
+    def _extend_drive(self, drive):
+        with self._timed_lock():
+            if not self._vm.isDomainReadyForCommands():
+                # Drive will be extended later by the periodic monitor.
+                self._log.warning("Domain not ready, retrying later")
+
+            if self._update_block_info([drive]):
+                self._handle_exceeded(drive, urgent=True)
 
     # Monitoring volumes.
 
     def monitor_volumes(self):
         """
-        Return True if at least one volume is being extended, False otherwise.
+        Check and extend drives if needed.
         """
-        drives = self._monitored_volumes()
-        if not drives:
-            return False
+        with self._timed_lock():
+            drives = self._monitored_volumes()
+            if not drives:
+                return
+
+            if not self._update_block_info(drives):
+                return
+
+            for drive in drives:
+                self._extend_drive_if_needed(drive)
+
+    def _update_block_info(self, drives):
+        """
+        Query libvirt block stats and update drives block info. This must be
+        done on every monitoring cycle, before we decide if a drive should be
+        extended.
 
+        Return True if the update was successful.
+        """
         try:
             block_stats = self._query_block_stats()
         except libvirt.libvirtError as e:
             self._log.error("Unable to get block stats: %s", e)
             return False
 
-        extended = False
         for drive in drives:
-            try:
-                if self._extend_drive_if_needed(drive, block_stats):
-                    extended = True
-            except ImprobableResizeRequestError:
-                break
+            self._query_block_info(drive, drive.volumeID, block_stats)
 
-        return extended
+        return True
 
-    def _extend_drive_if_needed(self, drive, block_stats):
+    def _extend_drive_if_needed(self, drive):
         """
         Check if a drive should be extended, and start extension flow if
         needed.
-
-        When libvirt BLOCK_THRESHOLD event handling is enabled (
-        irs.enable_block_threshold_event == True), this method acts according
-        the drive.threshold_state:
-
-        - UNSET: the drive needs to register for a new block threshold,
-                 so try to set it. We set the threshold both for chunked
-                 drives and non-chunked drives replicating to chunked
-                 drives.
-        - EXCEEDED: the drive needs extension, try to extend it.
-        - SET: this method should never receive a drive in this state,
-               emit warning and exit.
-
-        Return True if started an extension flow, False otherwise.
         """
-        if drive.threshold_state == storage.BLOCK_THRESHOLD.SET:
+        if drive.threshold_state == storage.BLOCK_THRESHOLD.UNSET:
+            self._handle_unset(drive)
```

Comment:
Will this handle the case if Vdsm is not running when the drive is not exceeded (i.e. exceeded drives during recovery)?

Reply:
Yes, in recovery we start with the drive in UNSET state, so in the first monitoring cycle we will check if the drive needs extension or set a block threshold.